### PR TITLE
Remove provider_can_use usages

### PR DESCRIPTION
**Pull request by josemigallas**
First part of [THREESCALE-3759: Remove (deprecated) non-APIAP code](https://issues.redhat.com/browse/THREESCALE-3759).

Removes all uses of `provider_can_use?(:api_as_product)`.

### CODE REVIEW

File: app/controllers/api/integrations_controller.rb [hunk 1]
```diff
@@ -69,9 +69,9 @@ def update_production
   def update_onpremises_production
     if @proxy.update_attributes(proxy_params)
       flash[:notice] = flash_message(:update_onpremises_production_success)
-      redirect_to action: :edit, anchor: 'production'
+      redirect_to action: :show, anchor: 'production'
     else
-      render :edit
+      render :show
     end
   end
```
Comment (Martouta) [inline app/controllers/api/integrations_controller.rb:76]: Please do not change this method in this PR and leave it for https://github.com/3scale/porta/pull/2288

File: test/integration/api/integrations_controller_test.rb [hunk 2]
```diff
@@ -173,12 +172,13 @@ def test_update_proxy_rule_position
...
-  def test_edit
+  # Regression test for pre-APIAP
+  def test_edit_not_found
     get edit_admin_service_integration_path(service_id: 'no-such-service')
     assert_response :not_found
 
     get edit_admin_service_integration_path(service_id: service.id)
-    assert_response :success
+    assert_response :not_found
   end
```
Comment (Martouta) [inline test/integration/api/integrations_controller_test.rb:182]: This test is completely removed in https://github.com/3scale/porta/pull/2288, plz do not change it here as well 😄

File: .rubocop.yml [hunk 3]
```diff
@@ -68,6 +68,8 @@ Metrics/MethodLength:
 
 Metrics/ClassLength:
   Max: 200
+  Exclude:
+    - 'test/**/*.rb'
```
Comment (Martouta) [inline .rubocop.yml:72]: Ok with this change but plz extract it to its own PR

File: app/controllers/api/integrations_controller.rb [hunk 4]
```diff
@@ -227,14 +227,31 @@ def proxy_rules_attributes
...
+      :error_no_match,
+      :error_status_limits_exceeded,
+      :error_headers_limits_exceeded,
+      :error_limits_exceeded,
+      :api_test_path,
+      :policies_config,
+      proxy_rules_attributes: %i[_destroy id http_method pattern delta metric_id redirect_url position last],
+      oidc_configuration_attributes: OIDCConfiguration::Config::ATTRIBUTES + [:id],
+      backend_api_configs_attributes: %i[_destroy id path]
     ]
```
Comment (Martouta) [inline app/controllers/api/integrations_controller.rb:255]: Way cleaner than before 👏🏼 🚀

File: app/controllers/api/services_controller.rb [hunk 5]
```diff
@@ -128,6 +128,7 @@ def can_edit_endpoints?
 
   # This will be the default 'settings' when apiap is live
   def settings_page
+    # TODO: remove all :settings associated pages
```
Comment (Martouta) [inline app/controllers/api/services_controller.rb:131]: It helps for later to search if all the `TODO` related to this issue start with `# TODO: THREESCALE-3759` as you have done in other PRs 😄

File: app/models/account.rb [hunk 6]
```diff
@@ -562,8 +562,9 @@ def provider_id_for_audits
     end
   end
 
+  # TODO: THREESCALE-3759 remove this method
   def independent_mapping_rules_enabled?
-    provider_can_use?(:api_as_product) || provider_can_use?(:independent_mapping_rules)
+    true
   end
```
Comment (Martouta) [inline app/models/account.rb:568]: You can directly remove it in https://github.com/3scale/porta/pull/2274#issuecomment-739999173 and perhaps that PR point to `master` directly

File: app/services/service_creator.rb [hunk 7]
```diff
@@ -37,12 +37,8 @@ def save!(params)
...
   def save_default_backend_api(params)
-    if provider_can_use?(:api_as_product)
-      return true if %i[path private_endpoint].none? { |key| params.key?(key) }
-      backend_api_proxy.update!(params)
-    else
-      backend_api_proxy.update!(private_endpoint: BackendApi.default_api_backend)
-    end
+    return true if %i[path private_endpoint].none? { |key| params.key?(key) }
+    backend_api_proxy.update!(params)
   end
```
Comment (Martouta) [inline app/services/service_creator.rb:42]: check https://github.com/3scale/porta/pull/2294

File: app/views/api/services/show.html.slim [hunk 8]
```diff
@@ -161,7 +161,6 @@ section class="service-widget"
           - [:buyers_manage_keys, :buyers_manage_apps, :buyer_plan_change_permission, :buyer_can_select_plan ].map { |setting| friendly_service_setting(service, setting) }.compact.each do |setting|
             li.item
               = setting
-          - if provider_can_use?(:api_as_product)
             li.item
```
Comment (Martouta) [inline app/views/api/services/show.html.slim:164]: This is not well aligned. Right not it is inside the `map` of setting above 😄 
It should be:
```ruby
          li.item
            ' ID for API calls is
            strong
              => service.id
            ' and system name is
            strong
              = service.system_name

          - [:buyers_manage_keys, :buyers_manage_apps, :buyer_plan_change_permission, :buyer_can_select_plan ].map { |setting| friendly_service_setting(service, setting) }.compact.each do |setting|
            li.item
              = setting
          li.item
            strong> Backends
            ' used in this product:
            ul
              - service.backend_apis.each do |backend_api|
                li = link_to backend_api.name, provider_admin_backend_api_path(backend_api)
```

File: features/support/paths.rb [hunk 9]
```diff
@@ -609,8 +611,11 @@ def path_to(page_name, *args)
     when /^the integration show page for (service ".+?")/
       admin_service_integration_path(Transform $1)
     when /^the integration page for (service ".+?")/
+      # TODO: THREESCALE-3759 edit page no longer exist, remove and place for show
```
Comment (Martouta) [inline features/support/paths.rb:614]: Just remove it in https://github.com/3scale/porta/pull/2288#discussion_r537506820 and do not replace it for anything. It just shouldn't be accessible and shouldn't exist. If there is any test pointing to it, we have to remove it or fix it depending on the test.

File: features/support/paths.rb [hunk 10]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+# rubocop:disable Metrics/MethodLength, Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
```
Comment (Martouta) [inline features/support/paths.rb:3]: In general, when you disable rubocop stuff, remember to enable it again at the end of the class/method/block/file.
That being said, I do not think that we should merge this change.

File: test/integration/api/integrations_controller_test.rb [hunk 11]
```diff
@@ -158,7 +158,6 @@ def test_update_proxy_rule_position
 
   test 'deploy is never called when saving proxy info for proxy pro users' do
     rolling_updates_on
-    Account.any_instance.stubs(:provider_can_use?).with(:api_as_product).returns(false)
```
Comment (Martouta) [inline test/integration/api/integrations_controller_test.rb:161]: This change belongs to https://github.com/3scale/porta/pull/2288

File: test/integration/api/integrations_controller_test.rb [hunk 12]
```diff
@@ -271,14 +271,6 @@ def test_edit
     assert_response :not_found
   end
 
-  test 'edit not found for apiap' do
-    rolling_updates_on
-    Account.any_instance.expects(:provider_can_use?).with(:api_as_product).returns(true)
-
-    get edit_admin_service_integration_path(service_id: service.id)
-    assert_response :not_found
-  end
```
Comment (Martouta) [inline test/integration/api/integrations_controller_test.rb:280]: This change belongs to https://github.com/3scale/porta/pull/2288

File: test/unit/signup/account_manager_test.rb [hunk 13]
```diff
@@ -127,7 +127,6 @@ class ProviderAccountManagerTest < Signup::AccountManagerTest
 
       test 'first service has a private endpoint in order to be functional for non-APIAP accounts' do
         Account.any_instance.stubs(:provider_can_use?).returns(false)
-        Account.any_instance.stubs(:provider_can_use?).with(:api_as_product).returns(false)
```
Comment (Martouta) [inline test/unit/signup/account_manager_test.rb:130]: check https://github.com/3scale/porta/pull/2294

**Review by Martouta — CHANGES_REQUESTED** [review]
This PR is huge and I couldn't review every detail. However, I could review big part of it.
I left some comments.
This is mostly a WIP with stuff to be extracted to their own PRs by logic units 😄 
Great job tho, all this is huge! 💪🏼 🚀

File: test/integration/api/integrations_controller_test.rb (hunk 2 above)
Comment (josemigallas) [reply test/integration/api/integrations_controller_test.rb:182]: It will break the build otherwise 🤷‍♂️

File: app/views/api/services/show.html.slim (hunk 8 above)
Comment (josemigallas) [reply app/views/api/services/show.html.slim:167]: Good catch!

File: features/support/paths.rb (hunk 9 above)
Comment (josemigallas) [reply features/support/paths.rb:614]: Yep, that's what the comment meant

File: features/support/paths.rb (hunk 10 above)
Comment (josemigallas) [reply features/support/paths.rb:3]: Yes I know... but I think it's silly rubocop don't have a page level disabling feature 😩

File: app/controllers/api/integrations_controller.rb [hunk 14]
```diff
@@ -228,14 +230,31 @@ def proxy_rules_attributes
...
+      :error_status_no_match,
+      :error_headers_no_match,
+      :error_no_match,
+      :error_status_limits_exceeded,
+      :error_headers_limits_exceeded,
+      :error_limits_exceeded,
+      :api_test_path,
+      :policies_config,
+      proxy_rules_attributes: %i[_destroy id http_method pattern delta metric_id redirect_url position last],
+      oidc_configuration_attributes: OIDCConfiguration::Config::ATTRIBUTES + [:id],
```
Comment (Martouta) [inline app/controllers/api/integrations_controller.rb:256]: rebase from master plz 🙈

Comment (josemigallas) [reply app/controllers/api/integrations_controller.rb:256]: It is up to date with master, what's wrong with this? 🤔

Comment (Martouta) [reply app/controllers/api/integrations_controller.rb:256]: right, my bad 👍🏼

File: features/support/paths.rb [hunk 15]
```diff
@@ -609,8 +609,11 @@ def path_to(page_name, *args)
     when /^the integration show page for (service ".+?")/
       admin_service_integration_path(Transform $1)
     when /^the integration page for (service ".+?")/
+      # TODO: THREESCALE-3759 edit page no longer exist, remove and replace for show
       edit_admin_service_integration_path(Transform $1)
     when 'the service integration page'
+      # TODO: THREESCALE-3759 edit page no longer exist, change for
+      # settings_admin_service_path(provider_first_service!)
```
Comment (Martouta) [inline features/support/paths.rb:616]:
```suggestion
      # TODO: THREESCALE-3759 edit page no longer exist, change
```

File: app/controllers/api/integrations_controller.rb [hunk 16]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+# rubocop:disable Metrics/ClassLength
```
Comment (Martouta) [inline app/controllers/api/integrations_controller.rb:3]: Rubocop class length in this case should be fixed in a following PR instead of permanently hiding the issue with this comment

Comment (Martouta) [reply app/controllers/api/integrations_controller.rb:3]: because once we remove the `edit` completely, this will be easy to fix 😉

File: app/controllers/api/integrations_controller.rb [hunk 17]
```diff
@@ -216,36 +220,49 @@ def proxy_rules_attributes
...
-      :error_headers_no_match, :error_no_match, :error_status_limits_exceeded, :error_headers_limits_exceeded, :error_limits_exceeded,
-      :api_test_path, :policies_config, proxy_rules_attributes: %i[_destroy id http_method pattern delta metric_id
-                                                                   redirect_url position last], oidc_configuration_attributes: OIDCConfiguration::Config::ATTRIBUTES + [:id]
-    ]
-
-    if Rails.application.config.three_scale.apicast_custom_url || @proxy.saas_configuration_driven_apicast_self_managed?
-      basic_fields << :endpoint
-      basic_fields << :sandbox_endpoint
-    end
+    permitted_fields = PROXY_BASIC_PARAMS
```
Comment (Martouta) [inline app/controllers/api/integrations_controller.rb:253]:
```suggestion
    permitted_fields = PROXY_BASIC_PARAMS.dup
```

File: app/controllers/api/integrations_controller.rb [hunk 18]
```diff
@@ -216,36 +220,49 @@ def proxy_rules_attributes
...
     if provider_can_use?(:apicast_oidc)
-      basic_fields << :oidc_issuer_endpoint
-      basic_fields << :oidc_issuer_type
-      basic_fields << :jwt_claim_with_client_id
-      basic_fields << :jwt_claim_with_client_id_type
+      permitted_fields << :oidc_issuer_endpoint
+      permitted_fields << :oidc_issuer_type
+      permitted_fields << :jwt_claim_with_client_id
+      permitted_fields << :jwt_claim_with_client_id_type
     end
```
Comment (Martouta) [inline app/controllers/api/integrations_controller.rb:263]:
```suggestion
    permitted_fields.merge!(%i[oidc_issuer_endpoint oidc_issuer_type jwt_claim_with_client_id jwt_claim_with_client_id_type]) if provider_can_use?(:apicast_oidc)
```

File: features/support/paths.rb (hunk 15 above)
Comment (josemigallas) [reply features/support/paths.rb:616]: Y U NO like this comment? Is it wrong?

File: app/controllers/api/integrations_controller.rb (hunk 16 above)
Comment (josemigallas) [reply app/controllers/api/integrations_controller.rb:3]: I think I'm just going to disable my rubocop integration!

File: app/controllers/api/integrations_controller.rb (hunk 18 above)
Comment (Martouta) [reply app/controllers/api/integrations_controller.rb:263]: in fact, we might not even need these at all after https://github.com/3scale/porta/pull/2288, but not sure

File: features/support/paths.rb (hunk 15 above)
Comment (Martouta) [reply features/support/paths.rb:616]: the thing is that this `the service integration page` is not used anymore from anywhere, so why replace it for X when we can just remove it because it is simply not used at all? 😄

File: app/controllers/api/integrations_controller.rb (hunk 16 above)
Comment (Martouta) [reply app/controllers/api/integrations_controller.rb:3]: hahahaah if that is the reason, you can have locally your own `.rubocop.yml` with your own rules. Ignored by `.git`

File: features/support/paths.rb (hunk 15 above)
Comment (josemigallas) [reply features/support/paths.rb:616]: ahh true, then it should be removed altogether.

**Review by Martouta — DISMISSED** [review]
Good job 👏🏼

File: test/integration/api/integrations_controller_test.rb (hunk 11 above)
Comment (josemigallas) [reply test/integration/api/integrations_controller_test.rb:159]: But this PR is supposed to remove all `provider_can_usse` usages 🤷‍♂️  even if this test will be removed, there's this smaller step first

**Review by Martouta — APPROVED** [review]
